compiler/rustc_hir_analysis/src/structured_errors/missing_cast_for_variadic_arg.rs

```diff
@@ -21,28 +21,28 @@ impl<'tcx> StructuredDiagnostic<'tcx> for MissingCastForVariadicArg<'tcx, '_> {
     }
 
     fn diagnostic_common(&self) -> DiagnosticBuilder<'tcx, ErrorGuaranteed> {
-        let mut err = self.sess.struct_span_err_with_code(
-            self.span,
-            &format!("can't pass `{}` to variadic function", self.ty),
-            self.code(),
-        );
+        let mut err_struct = errors::PassToVariadicFunction {
+            span: self.span,
+            ty: self.ty,
+            cast_ty: self.cast_ty,
+            help: None,
+            replace: "".to_string(),
+            sugg_span: None,
+        };
 
         if self.ty.references_error() {
+            let mut err = self.sess.create_err(err_struct.to_owned());
             err.downgrade_to_delayed_bug();
         }
 
         if let Ok(snippet) = self.sess.source_map().span_to_snippet(self.span) {
-            err.span_suggestion(
-                self.span,
-                &format!("cast the value to `{}`", self.cast_ty),
-                format!("{} as {}", snippet, self.cast_ty),
-                Applicability::MachineApplicable,
-            );
+            err_struct.sugg_span = Some(self.span);
+            err_struct.replace = format!("{} as {}", snippet, self.cast_ty);
         } else {
-            err.help(&format!("cast the value to `{}`", self.cast_ty));
+            err_struct.help = Some(())
```

Comment:
Instead of setting err_struct's values, move this computation earlier so you can just pass the right values to the struct constructor on line 24?